File: internal/state/state.go
package state

import (
	"bufio"
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"

	"github.com/cbaumont/nina/internal/llm"
)

const dirName = ".nina"

type Session struct {
	SessionID string         `json:"session_id"`
	Goal      string         `json:"goal"`
	State     string         `json:"state"`
	PlanTitle string         `json:"plan_title"`
	Steps     []llm.PlanStep `json:"steps"`
	StepIndex int            `json:"step_index"`
	Snapshots int            `json:"snapshots"`
	LastRef   string         `json:"last_ref"`
}

func Save(workspaceDir string, sess *Session, messages []llm.Message) error {
	dir := filepath.Join(workspaceDir, dirName)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	raw, err := json.MarshalIndent(sess, "", "  ")
	if err != nil {
		return err
	}
	if err := writeAtomic(filepath.Join(dir, "session.json"), raw); err != nil {
		return err
	}
	var transcript bytes.Buffer
	for _, message := range messages {
		line, err := json.Marshal(message)
		if err != nil {
			return err
		}
		transcript.Write(line)
		transcript.WriteByte('\n')
	}
	return writeAtomic(filepath.Join(dir, "transcript.jsonl"), transcript.Bytes())
}

func Load(workspaceDir string) (*Session, []llm.Message, error) {
	raw, err := os.ReadFile(filepath.Join(workspaceDir, dirName, "session.json"))
	if errors.Is(err, os.ErrNotExist) {
		return nil, nil, nil
	}
	if err != nil {
		return nil, nil, err
	}
	sess := &Session{}
	if err := json.Unmarshal(raw, sess); err != nil {
		return nil, nil, fmt.Errorf("reading session.json: %w", err)
	}
	messages, err := loadTranscript(filepath.Join(workspaceDir, dirName, "transcript.jsonl"))
	if err != nil {
		return nil, nil, err
	}
	return sess, messages, nil
}

func loadTranscript(path string) ([]llm.Message, error) {
	file, err := os.Open(path)
	if errors.Is(err, os.ErrNotExist) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var messages []llm.Message
	scanner := bufio.NewScanner(file)
	scanner.Buffer(nil, 4*1024*1024)
	for scanner.Scan() {
		line := bytes.TrimSpace(scanner.Bytes())
		if len(line) == 0 {
			continue
		}
		var message llm.Message
		if err := json.Unmarshal(line, &message); err != nil {
			return nil, fmt.Errorf("reading transcript.jsonl: %w", err)
		}
		messages = append(messages, message)
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return messages, nil
}

func SaveHistory(workspaceDir, text string) error {
	dir := filepath.Join(workspaceDir, dirName)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	return writeAtomic(filepath.Join(dir, "history.md"), []byte(text))
}

func LoadHistory(workspaceDir string) (string, error) {
	raw, err := os.ReadFile(filepath.Join(workspaceDir, dirName, "history.md"))
	if errors.Is(err, os.ErrNotExist) {
		return "", nil
	}
	if err != nil {
		return "", err
	}
	return string(raw), nil
}

func writeAtomic(path string, data []byte) error {
	tmp := path + ".tmp"
	if err := os.WriteFile(tmp, data, 0o644); err != nil {
		return err
	}
	return os.Rename(tmp, path)
}

File: src/nina/system/state.py
from __future__ import annotations

import contextlib
import json
from dataclasses import asdict, dataclass, field
from pathlib import Path

from nina.agent.tools import PlanStep

DIR_NAME = ".nina"


@dataclass
class Session:
    session_id: str = ""
    goal: str = ""
    state: str = ""
    plan_title: str = ""
    steps: list[PlanStep] = field(default_factory=list)
    step_index: int = 0
    snapshots: int = 0
    last_ref: str = ""
    sdk_session_id: str | None = None
    model: str | None = None


def save(workspace_dir: str, sess: Session) -> None:
    dir_path = Path(workspace_dir) / DIR_NAME
    dir_path.mkdir(parents=True, exist_ok=True)
    raw = json.dumps(asdict(sess), indent=2)
    _write_atomic(dir_path / "session.json", raw)


def load(workspace_dir: str) -> Session | None:
    path = Path(workspace_dir) / DIR_NAME / "session.json"
    if not path.exists():
        return None
    data = json.loads(path.read_text())
    steps = [PlanStep(title=s["title"], goal=s["goal"]) for s in data.get("steps", [])]
    return Session(
        session_id=data.get("session_id", ""),
        goal=data.get("goal", ""),
        state=data.get("state", ""),
        plan_title=data.get("plan_title", ""),
        steps=steps,
        step_index=data.get("step_index", 0),
        snapshots=data.get("snapshots", 0),
        last_ref=data.get("last_ref", ""),
        sdk_session_id=data.get("sdk_session_id"),
        model=data.get("model"),
    )


def save_history(workspace_dir: str, text: str) -> None:
    dir_path = Path(workspace_dir) / DIR_NAME
    dir_path.mkdir(parents=True, exist_ok=True)
    _write_atomic(dir_path / "history.md", text)


def load_history(workspace_dir: str) -> str:
    path = Path(workspace_dir) / DIR_NAME / "history.md"
    if not path.exists():
        return ""
    return path.read_text()


def append_transcript(workspace_dir: str, entry: dict[str, object]) -> None:
    dir_path = Path(workspace_dir) / DIR_NAME
    dir_path.mkdir(parents=True, exist_ok=True)
    path = dir_path / "transcript.jsonl"
    with path.open("a") as f:
        f.write(json.dumps(entry) + "\n")


def load_transcript(workspace_dir: str) -> list[dict[str, object]]:
    path = Path(workspace_dir) / DIR_NAME / "transcript.jsonl"
    if not path.exists():
        return []
    entries: list[dict[str, object]] = []
    for line in path.read_text().splitlines():
        line = line.strip()
        if not line:
            continue
        with contextlib.suppress(json.JSONDecodeError):
            entries.append(json.loads(line))
    return entries


def _write_atomic(path: Path, data: str) -> None:
    tmp = path.with_suffix(path.suffix + ".tmp")
    tmp.write_text(data)
    tmp.rename(path)
